Aarh third deliverable

AlanAndresRH opened the pull request.

AlanAndresRH committed:
First delivery Home module implementation

AlanAndresRH committed:
Second delivery Cryptocurrency app

AlanAndresRH committed:
Third delivery Cryptocurrency app

AlanAndresRH committed:
Third delivery Cryptocurrency app (configuration navController)

AlanAndresRH committed:
Completion of Room implementation in the entire project and implementation of navigation controller

AlanAndresRH committed:
Completion of Room implementation in the entire project and implementation of navigation controller

AlanAndresRH committed:
Completion of Room implementation in the entire project and implementation of navigation controller

AlanAndresRH committed:
Correction of observations in review of the first pull request

AlanAndresRH committed:
Correction of observations in review of the first pull request 2
HttpLoggingInterceptor is added and also the User-Agent header
Logic fix in AsksBookEntity and BidsBookEntity entity

AlanAndresRH committed:
Debugging warnings and errors

AlanAndresRH committed:
Init branch
Spaces are removed in vector .xml files

AlanAndresRH committed:
Implementation of network status check
Connection error layout
LocalDataSource is added

AlanAndresRH committed:
Resolution of observations of pull request #2
Added and initialized in terminal ktlint

AlanAndresRH committed:
Initialization of new and final branch

diego-parra-robayo commented on app/src/main/java/com/example/finalprojectwizelinecryptocurrencies/data/cryptocurrencies/dataSource/local/BookLocalDataSource.kt:
```diff
@@ -0,0 +1,47 @@
+        orderBook: OrderBookEntity,
+        asksBookEntity: List<AsksBookEntity>,
+        bidsBookEntity: List<BidsBookEntity>
+    ) {
```
Esto es un detalle adicional, pero para mantener la integridad de los datos en la base de datos, estos métodos deberían correr en una transacción.
Para generar una transacción en el dataSource, puedes inyectar la base de datos en el constructor, y luego usar database.withTransaction { ... delete ... insert ... }

AlanAndresRH replied:
Listo, observación realizada

diego-parra-robayo commented on app/src/main/java/com/example/finalprojectwizelinecryptocurrencies/ui/home/viewModel/HomeViewModel.kt:
```diff
@@ -0,0 +1,107 @@
+    fun changeFilterKeyRxJava(key: KeyFilter) {
+        _state.update { it.copy(isLoading = true) }
+
+        viewModelScope.launch {
```
Como esta implementación está en RxJava ya no es necesario usar viewModelScope.launch{}

AlanAndresRH replied:
Listo, observación realizada

diego-parra-robayo commented on app/src/main/java/com/example/finalprojectwizelinecryptocurrencies/ui/home/viewModel/HomeViewModel.kt:
```diff
@@ -0,0 +1,107 @@
+        _state.update { it.copy(isLoading = true) }
+
+        viewModelScope.launch {
+            val resBook = getBooksFilterUseCase.invokeRxJava(key)
```
Para cancelar los observables al eliminar el ViewModel y evitar memory leaks usa un CompositeDisposable.
Puedes crearlo como variable global, añadir cada observable a éste mediante el método disposable.add(_) y cerrarlo en el método onCleared del viewModel mediante disposable.dispose(_)

AlanAndresRH replied:
Listo, observación realizada

diego-parra-robayo commented on app/src/main/java/com/example/finalprojectwizelinecryptocurrencies/ui/state/HomeState.kt:
```diff
@@ -0,0 +1,14 @@
+    val books: List<Book> = emptyList(),
+    val isLoading: Boolean = false,
+    val keyFilter: KeyFilter = KeyFilter.FILTER_MXN,
+    val showMexico: Boolean = false,
```
showMexico siempre depende del valor de KeyFilter. En lugar de asignarse cada vez en el ViewModel podría colocarse una función dentro del state como showMexico() = keyFilter == KeyFilter.FILTER_MXN.

*El mismo caso sería para showAllCountry y showErrorData.

AlanAndresRH replied:
Listo, observación realizada

diego-parra-robayo commented on app/src/main/java/com/example/finalprojectwizelinecryptocurrencies/ui/state/HomeState.kt:
```diff
@@ -0,0 +1,14 @@
+package com.example.finalprojectwizelinecryptocurrencies.ui.state
```
Esta clase, HomeState, queda mejor dentro de la carpeta home. Del mismo modo, DetailState se puede colocar en la carpeta detail.

AlanAndresRH replied:
Listo, observación realizada

diego-parra-robayo reviewed (commented):
Muy buen trabajo. Te dejo un par de comentarios y quedo al pendiente de los unit tests.

AlanAndresRH committed:
Restructuring of HomeViewModel.
UnitTest HomeViewModel of Use Case.
Thread Module implementation.
Shimmer implementation.

AlanAndresRH committed:
Correction of observations in Pull request 3
Initialization of unit tests in the ui

AlanAndresRH committed:
Ui Test Implementation.
Separation of functions in GetBooksFilterUseCase.

AlanRH14 committed:
Merge pull request #4 from AlanRH14/master

Master
